ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/AvroDiffy.scala

```diff
@@ -18,118 +18,164 @@
-                )
-                .toMap
-              val r = y
-                .flatMap(outer =>
-                  Option(outer.get(name).asInstanceOf[java.util.List[GenericRecord]].asScala.toList)
-                )
-                .getOrElse(List())
-                .flatMap(inner =>
-                  Try(inner.get(unorderedFieldKeys(fullName))).toOption.map(k => (k, inner))
-                )
-                .toMap
-              (l.keySet ++ r.keySet).flatMap(k => diff(l.get(k), r.get(k), fullName)).toList
-            } else {
-              val a = x
-                .flatMap(r => Option(r.get(name).asInstanceOf[java.util.List[GenericRecord]]))
-                .map(sortList)
-              val b = y
-                .flatMap(r => Option(r.get(name).asInstanceOf[java.util.List[GenericRecord]]))
-                .map(sortList)
-              if (a == b) {
-                Nil
-              } else {
-                Seq(Delta(fullName, a, b, delta(a.orNull, b.orNull)))
-              }
-            }
-          case _ =>
-            val a = x.flatMap(r => Option(r.get(name)))
-            val b = y.flatMap(r => Option(r.get(name)))
-            if (a == b) Nil else Seq(Delta(fullName, a, b, delta(a.orNull, b.orNull)))
-        }
-      }
-      .filter(d => !ignore.contains(d.field))
+    data
+  }.recover { case _: NoSuchFieldException | _: IllegalAccessException =>
+    // Return default instance
+    SpecificData.get()
+  }.get
+
+  override def apply(x: T, y: T): Seq[Delta] = (x, y) match {
+    case (null, null)                    => Seq.empty
+    case (_, null)                       => Seq(Delta("", Some(x), None, UnknownDelta))
+    case (null, _)                       => Seq(Delta("", None, Some(y), UnknownDelta))
+    case _ if x.getSchema != y.getSchema => Seq(Delta("", Some(x), Some(y), UnknownDelta))
+    case _                               => diff(x, y, x.getSchema, "")
+  }
+
+  private def isRecord(schema: Schema): Boolean = schema.getType match {
+    case Schema.Type.RECORD => true
+    case Schema.Type.UNION  => schema.getTypes.asScala.map(_.getType).contains(Schema.Type.RECORD)
+    case _                  => false
+  }
+
+  private def isNumericType(`type`: Schema.Type): Boolean = `type` match {
+    case Schema.Type.INT | Schema.Type.LONG | Schema.Type.FLOAT | Schema.Type.DOUBLE => true
+    case _                                                                           => false
+  }
+
+  private def numericValue(value: AnyRef): Double = value match {
+    case i: java.lang.Integer => i.toDouble
+    case l: java.lang.Long    => l.toDouble
+    case f: java.lang.Float   => f.toDouble
+    case d: java.lang.Double  => d
+    case _ => throw new IllegalArgumentException(s"Unsupported numeric type: ${value.getClass}")
   }
 
-  private def getRawType(schema: Schema): Schema = {
-    schema.getType match {
+  private def diff(x: AnyRef, y: AnyRef, schema: Schema, field: String): Seq[Delta] = {
```

Comment:
would it be possible to get more specific types here than `AnyRef` for x and y?

Reply:
I don't think this is possible.
The goal here is to call this recursively and branch the diff according to the schema type information + casting.

We can potentially create dedicated diff methods like `diffRecord(x: IndexedRecord, y: IndexedRecord)` but this does not play well with parameterized types like `list` and `maps`

Reply:
ah, I see. I trust your judgement, and any type other than `AnyRef` we choose probably would have to be super generic anyway.